### DISCUSSION TITLE
Roadmap

### DISCUSSION THREAD

**Discussion by Hopding** [Hopding/pdf-lib#998] [Announcements]
# Planned
Stuff that will definitely happen and has a clear path forward.
- [ ] Outlines
  * https://github.com/Hopding/pdf-lib/pull/486
  * https://github.com/Hopding/pdf-lib/issues/567
  * https://github.com/Hopding/pdf-lib/issues/450
- [ ] Update `@pdf-lib/fontkit`
  * See if this fixes https://github.com/Hopding/pdf-lib/pull/726
  * https://github.com/Hopding/pdf-lib/issues/971
  * https://github.com/Hopding/pdf-lib/issues/944
  * https://github.com/Hopding/pdf-lib/issues/826
  * https://github.com/Hopding/pdf-lib/issues/808
  * https://github.com/Hopding/pdf-lib/issues/785
  * https://github.com/Hopding/pdf-lib/issues/746
  * https://github.com/Hopding/pdf-lib/issues/688
  * https://github.com/Hopding/pdf-lib/issues/675
  * https://github.com/Hopding/pdf-lib/issues/634
  * https://github.com/Hopding/pdf-lib/issues/231
- [ ] Encryption/Decryption:
  * https://github.com/Hopding/pdf-lib/issues/472
  * https://github.com/Hopding/pdf-lib/issues/443
  * https://github.com/Hopding/pdf-lib/issues/243
  * https://github.com/Hopding/pdf-lib/issues/233
- [ ] Annotations:
  * https://github.com/Hopding/pdf-lib/issues/878
- [ ] Reduce bundle size:
  * https://github.com/Hopding/pdf-lib/issues/566
  
# Unplanned
Stuff that might happen but is going to take longer.
- [ ] Replace template text
  * https://github.com/Hopding/pdf-lib/issues/177
- [ ] Reduce memory usage by streaming i/o instead of reading everything into memory at once:
  * https://github.com/Hopding/pdf-lib/issues/990
  * https://github.com/Hopding/pdf-lib/issues/197
  * https://github.com/Hopding/pdf-lib/issues/121
- [ ] Additional example documentation:
  * https://github.com/Hopding/pdf-lib/issues/973
- [ ] Improve typography and text layout support:
  * https://github.com/Hopding/pdf-lib/issues/972
  * https://github.com/Hopding/pdf-lib/issues/970
  * https://github.com/Hopding/pdf-lib/issues/865
  * https://github.com/Hopding/pdf-lib/issues/853
  * https://github.com/Hopding/pdf-lib/issues/774
  * https://github.com/Hopding/pdf-lib/issues/763
  * https://github.com/Hopding/pdf-lib/issues/752
  * https://github.com/Hopding/pdf-lib/issues/657
  * https://github.com/Hopding/pdf-lib/issues/588
  * https://github.com/Hopding/pdf-lib/issues/578
  * https://github.com/Hopding/pdf-lib/issues/930
  * https://github.com/Hopding/pdf-lib/issues/935
- [ ] Image listing/extraction:
  * https://github.com/Hopding/pdf-lib/issues/83
  * https://github.com/Hopding/pdf-lib/issues/963
- [ ] Framework-specific examples:
  * https://github.com/Hopding/pdf-lib/issues/941
  * https://github.com/Hopding/pdf-lib/issues/749
- [ ] Tagged PDF support:
  * https://github.com/Hopding/pdf-lib/issues/895
- [ ] Page labels:
  * https://github.com/Hopding/pdf-lib/issues/464
  * https://github.com/Hopding/pdf-lib/issues/878
- [ ] Better documentation about fonts:
  * https://github.com/Hopding/pdf-lib/issues/871
- [ ] Incremental updates:
  * https://github.com/Hopding/pdf-lib/issues/834
  * https://github.com/Hopding/pdf-lib/issues/816
- [ ] Improve behavior & docs around width/height/sizing and boxes:
  * https://github.com/Hopding/pdf-lib/issues/764
  * https://github.com/Hopding/pdf-lib/issues/818
- [ ] Set acrofield values by index:
  * https://github.com/Hopding/pdf-lib/issues/811
- [ ] Set preferred font for acro text field:
  * https://github.com/Hopding/pdf-lib/issues/674
- [ ] Override field line height & font size to make text fit:
  * https://github.com/Hopding/pdf-lib/issues/974
- [ ] Links & named destinations:
  * https://github.com/Hopding/pdf-lib/issues/622
  * https://github.com/Hopding/pdf-lib/issues/858
- [ ] Page cloning:
  * https://github.com/Hopding/pdf-lib/issues/169
- [ ] Page deletion and object removal:
  * https://github.com/Hopding/pdf-lib/issues/140
- [ ] Additional SVG options:
  * https://github.com/Hopding/pdf-lib/issues/738
  * https://github.com/Hopding/pdf-lib/issues/739
- [ ] Fix `drawPage` opacity option for embedded pages:
  * https://github.com/Hopding/pdf-lib/issues/553
  
# TBD
Stuff that might be useful but needs research.
- [ ] Create/modify large documents - this may be distinct from i/o streaming in that it requires the ability to make large numbers of changes before deciding to finally serialize the document. A potential solution might simply be incremental updates. Users could simply "flush" the document periodically to avoid building up too much data in memory. Another (more complicated) solution might be to serialize objects in temp buffers/files until they need to be loaded in memory. This would likely add significant complexity though, so the incremental update solution is more appealing (especially for a first pass). 
  Could supplement the incremental update approach with the ability to squash a document with incremental updates into a single "update", thereby removing the incremental updates.
  * https://github.com/Hopding/pdf-lib/issues/990
  * https://github.com/Hopding/pdf-lib/issues/197
  * https://github.com/Hopding/pdf-lib/issues/121
- [ ] Transparency Groups
- [ ] Optional Content Groups
  * https://github.com/Hopding/pdf-lib/issues/907
- [ ] Publish to deno.land:
  * https://github.com/Hopding/pdf-lib/issues/837
- [ ] XMP and custom metadata:
  * https://github.com/Hopding/pdf-lib/issues/807
  * https://github.com/Hopding/pdf-lib/issues/55
- [ ] Better extract/copy/merge/split API:
  * https://github.com/Hopding/pdf-lib/issues/806
  * https://github.com/Hopding/pdf-lib/issues/567
  * https://github.com/Hopding/pdf-lib/issues/341
  * https://github.com/Hopding/pdf-lib/issues/218
- [ ] Digital signature support:
  * https://github.com/Hopding/pdf-lib/issues/794
  * https://github.com/Hopding/pdf-lib/issues/371
  * https://github.com/Hopding/pdf-lib/issues/172
- [ ] Sanitization support:
  * https://github.com/Hopding/pdf-lib/issues/433
  * https://github.com/Hopding/pdf-lib/issues/790
- [ ] PDF/A support:
  * https://github.com/Hopding/pdf-lib/issues/230
  * https://github.com/Hopding/pdf-lib/issues/756
  * https://www.pdfa.org/wp-content/uploads/2015/12/StructureElementsBestPracticeGuide_2016-01-19.pdf
- [ ] Update compiler target?
  * https://github.com/Hopding/pdf-lib/issues/747
- [ ] Reset (rotated) page's origin:
  * https://github.com/Hopding/pdf-lib/issues/587
  * https://github.com/Hopding/pdf-lib/issues/545
- [ ] API to extract graphic objects:
  * https://github.com/Hopding/pdf-lib/issues/581
- [ ] Acrofield widget documentation/examples:
  * https://github.com/Hopding/pdf-lib/issues/602
- [ ] Table support:
  * https://github.com/Hopding/pdf-lib/issues/382
- [ ] Linearization support

**Comment by sepsol** [comment]
Thanks for sharing the roadmap and your hard work. Would love to see the `digital signature` feature happen in the near future!

**Reply by joshmarinacci** [reply]
If I wanted to start working on text editing support (or at least searching for all text nodes in a doc) where would I start? (I'm a full time JS/TS engineer)

**Comment by Dude29** [comment]
Thank you for your incredible work!

**Comment by MickL** [comment]
@Hopding All issues of planned and unplanned are closed as completed. Are they all actually resolved?

**Comment by TheOneTrueRy** [comment]
What's the point of a roadmap or issue/feature "tracker" when everything gets immediately closed as completed without actually being worked on or ever implemented lmao
